File: ContestOfSharp/ContestOfSharp/Challenges/Challenge12.cs
using BankaiSouth;
namespace ContestOfSharp.Challenges;

/*
 * Crea una función que reciba dos cadenas como parámetro (str1, str2)
 * e imprima otras dos cadenas como salida (out1, out2).
 * - out1 contendrá todos los caracteres presentes en la str1 pero NO
 *   estén presentes en str2.
 * - out2 contendrá todos los caracteres presentes en la str2 pero NO
 *   estén presentes en str1.
 */
public abstract class Challenge12
{
    public static (string, string) OuterRim(string str1, string str2)
    {
        HashSet<char> set1 = [..str1.ToCharArray()];
        HashSet<char> set2 = [..str2.ToCharArray()];
        
        var commonCharacters = set1.Intersect(set2).ToArray();
        var inFirstButNotInSecond = set1.Except(commonCharacters);
        var inSecondButNotInFirst = set2.Except(commonCharacters);

        var result = (
            inFirstButNotInSecond.ToArray().Joined(""),
            inSecondButNotInFirst.ToArray().Joined("")
        ); 
        
        Console.WriteLine($"out1: {result.Item1}");
        Console.WriteLine($"out2: {result.Item2}");

        return result;
    }
}

File: ContestOfSwift/ContestOfSwift/Challenges/Challenge12.swift
//
//  Challenge12.swift
//  ContestOfSwift
//
//  Created by Sergio Daniel on 4/02/25.
//

/*
 * Crea una función que reciba dos cadenas como parámetro (str1, str2)
 * e imprima otras dos cadenas como salida (out1, out2).
 * - out1 contendrá todos los caracteres presentes en la str1 pero NO
 *   estén presentes en str2.
 * - out2 contendrá todos los caracteres presentes en la str2 pero NO
 *   estén presentes en str1.
 */

enum Challenge12 {
    
    static func outerRim(str1: String, str2: String) -> (out1: String, out2: String) {
        let set1 = Set(str1)
        let set2 = Set(str2)
        
        let sub1 = Array(set1.subtracting(set2))
        let sub2 = Array(set2.subtracting(set1))
        
        return (
            out1: str1.filter { sub1.contains(String($0)) }
                      .map(String.init)
                      .joined(separator: ""),
            out2: str2.filter { sub2.contains(String($0)) }
                      .map(String.init)
                      .joined(separator: "")
        )
    }
    
}
